### ISSUE TITLE
Angular 17 Prerendering (Static Site Generation (SSG)) does not wait until the request is done when using with Firebase

### ISSUE THREAD

**Issue by pungme** [angular/angular#54267]
### Which @angular/* package(s) are the source of the bug?

platform-server

### Is this a regression?

Yes

### Description

Hi

I'm pre-rendering an angular site with 100+ pages and each page has a simple call to Firebase to get data and then display them on the page. I put all routes in routes.txt

 "prerender": {
   "routesFile": "routes.txt"
 },

It worked well until I realized that some of pre-render pages had empty content. Every time I run the prerender, it has random results, some pages works some pages don't. if I only have 2-5 routes, the pre-render works fine and consistent. My suspicious is that when pre-render, angular just blasts calls to Firebase and doesn't wait until Firebase returns the data, hence the empty pre-render page.

Is there a way to set pre-render to do it synchronously or a way for prerender to wait until each page is finished?

Thank you so much in advance

### Please provide the exception or error you saw

```true
no error
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.1.2
Node: 20.11.0
Package Manager: npm 10.2.4
OS: darwin arm64

Angular: 17.1.2
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router, ssr

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1701.2
@angular-devkit/build-angular   17.1.2
@angular-devkit/core            17.1.2
@angular-devkit/schematics      17.1.2
@angular/fire                   17.0.1
@schematics/angular             17.1.2
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2
```

**Comment by alan-agius4** [comment]
Can you please provide the code that you are using the make the request?

**Comment by pungme** [comment]
Hi Alan, 

Thank you so much for a prompt reply. Attached you will find the minimal reproducible code. just run ng build and the problem should occur. 

[allchords.zip](https://github.com/angular/angular/files/14178410/allchords.zip)

In routes.txt, there are 11 routes there. When I run ´ng build´ it creates prerender in dist/allchords/browser of all 11 but not all of them have content inside. 

<img width="1288" alt="image" src="https://github.com/angular/angular/assets/1815907/74945b57-6f7b-4217-ad17-822cbc975311">

But if I run it again and again, sometimes it works with the same route. Or to prove it, just remove all other routes and only have one, then the prerender created correctly with content 

<img width="1462" alt="image" src="https://github.com/angular/angular/assets/1815907/bcfe12e4-b4a9-424a-9516-7128ab16fc05">

I also noticed that shuffling around the order of route makes a difference in what gets to render successfully but I could not pin down what is the pattern here. Hence I suspect that this is either a race condition or pre-render don't wait for the. 

Thanks again and let me know if you need anything else from my side

Pung

**Comment by imaksp** [comment]
Hi @pungme , don't know for how much time angular pre render process waits for the data from firebase , but to solve this you can use ResolveFn in routes file, with it you can directly get data in component using this.activatedRoute.data.subscribe fn.

**Comment by pungme** [comment]
Hi Akash, I am not sure what you meant by that. Do you have any example that I can follow? Thanks!

**Comment by imaksp** [comment]
Hi, adding one more thing, you are using following code in ngOnInit & using ngIf in template, in this angular has no proper way to find (during prerender) whether data loading is completed or not:
```ts
 this.route.params.subscribe(params => {
      this.firestore
      .collection('songs', ref => 
        ref.where('url_id', '==', params["chordid"])
           .where('isPublished', '==', true)
           .limit(1)
      )
      .valueChanges()
      .subscribe((data: any[]) => {
        // console.log(data[0].chords)
        if (data.length > 0) {
          this.songObject = data[0];
      }
}
```

So I think Angular should add support for new function like [getStaticProps](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props) from Next.js or ask user to use promise based resolveFn in route definition. 

this way angular can make sure that data is loaded before prerender process for page starts.

**Comment by msacket** [comment]
I ran into this problem.  I didn't like the resolver function because it stops navigation during normal routing.  What I did instead was write a callable function that gives me the results.  Angular waits for these during prerendering or ssr.

```
// firebase function
export const getSongsForChordId = functions.https.onCall((data, context) => {
  const {chordid} = data;
  return db.collection("songs").where("url_id", "==" chordid)
    .limit(20)
    .get();
});
```

```
// service
songsForChordId$(chordid: string): Observable<Song[]> {
    const SONG_KEY = makeStateKey<any>(`${chordid}-songs`);

    if (this.transferState.hasKey(STATE_KEY)) {
      const data = this.transferState.get<any>(SONG_KEY, null);
      return of(data);
    }
    return httpsCallableData<GetSongsParams, Song[]>(
      this.fn,
      'getSongsForChordId'
    )({ chordid }).pipe(
      tap((songs) => {
        this.transferState.set(SONG_KEY, songs);
      })
    );
  }
``

**Comment by pungme** [comment]
Hi,

Thank you so much for the suggestion! 

Another observation I found that the first 4 routes always pre-rendered successfully. Is this an expected behaviour? 

@msacket @imaksp Thank you so much for your help! 🙇🏻‍♂️ I tried the suggested solutions, but I'm having a bit of trouble wrapping my head around the current code and the new approach you suggested. Would you please be so kind as to provide a more descriptive explanation of which code I would need to change and so on?

Thank you very much,
Pung

**Comment by pungme** [comment]
@subhamoy-ghosh  Great to hear that I'm not alone. I will try to downgrade to Angular 16 and see if the problem is fixed

**Comment by msacket** [comment]
@pungme I don't believe a downgrade will fix your problem.  I believe Angular is or was intercepting http requests in `zone.js` to wait for them to complete.   It appears there is a different method now see (https://dev.to/krisplatis/angular-ssr-v16-saying-goodbye-to-a-sneaky-trick-macrotask-wrapping-for-http-calls-487b). 

Firestore is using websockets to push real-time updates.  That's why you'd need to use a callable function (http request) to fetch the data instead.  Angular SSR handles those just fine.

**Comment by imaksp** [comment]
Hello all, just checked the above article it seems Angular is tracking all http requests using interceptor, but I believe this solution should only be used for SSR. & its overkill for simple SSG like use case.

for SSG (& SSG with ISR) simple solution like Next.js's   [getStaticProps](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props) should be supported, with this approach it will be easier to debug. 
Also this function will [only run on server](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props#when-does-getstaticprops-run) (so more clear separation) & we can safely use secret keys to fetch static data from CMS.

@eneajaho what's your though on this ?

**Comment by davidbusuttil** [comment]
Same issue here... using ResolveFn is not a pretty solution... any other suggestions?

**Comment by imaksp** [comment]
Hello all, it seems there is new more explicit way for fetching such data is coming in v18 (using ExperimentalPendingTasks):

Related PR: https://github.com/angular/angular/pull/55487

With this Angular will wait for all async tasks (normal http request/ websockets calls) to complete before marking app/component as stable (before prerender).
but I am not sure if its possible to make it run on server only (for using secret key safely).

**Comment by JeanMeche** [comment]
The general idea, is yes you need to keep your App unstable to hold off the generation and wait for your data. PendingTasks is the dedicated service for that. 

https://angular.dev/api/core/PendingTasks
